core/src/main/java/com/linecorp/armeria/common/CompositeHttpHeadersBase.java

```diff
@@ -0,0 +1,224 @@
+ * to avoid expensive copy operations.
+ */
+class CompositeHttpHeadersBase
+        extends CompositeStringMultimap</* IN_NAME */ CharSequence, /* NAME */ AsciiString>
+        implements HttpHeaderGetters {
+
+    private static final Supplier<StringMultimap<CharSequence, AsciiString>> DEFAULT_APPENDER_SUPPLIER =
+            () -> new HttpHeadersBase(DEFAULT_SIZE_HINT);
+
+    CompositeHttpHeadersBase(HttpHeaderGetters... parents) {
+        super(from(parents), DEFAULT_APPENDER_SUPPLIER);
+    }
+
+    CompositeHttpHeadersBase(@Nullable CompositeStringMultimap<CharSequence, AsciiString> additionals,
+                             List<HttpHeaderGetters> parents,
+                             @Nullable CompositeStringMultimap<CharSequence, AsciiString> defaults) {
+        super(additionals, from(parents), defaults, DEFAULT_APPENDER_SUPPLIER);
+    }
+
+    private static List<StringMultimap<CharSequence, AsciiString>> from(HttpHeaderGetters... headers) {
+        if (headers.length == 0) {
+            return Collections.emptyList();
+        }
+
+        final ImmutableList.Builder<StringMultimap<CharSequence, AsciiString>> builder =
+                ImmutableList.builder();
+        for (HttpHeaderGetters header : headers) {
+            if (header instanceof HttpHeadersBase) {
+                builder.add((HttpHeadersBase) header);
+            } else {
+                builder.add(new HttpHeadersBase(header));
+            }
+        }
+
+        return builder.build();
+    }
+
+    private static List<StringMultimap<CharSequence, AsciiString>> from(List<HttpHeaderGetters> headers) {
+        if (headers.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        final ImmutableList.Builder<StringMultimap<CharSequence, AsciiString>> builder =
+                ImmutableList.builder();
+        for (HttpHeaderGetters header : headers) {
+            if (header instanceof HttpHeadersBase) {
+                builder.add((HttpHeadersBase) header);
+            } else {
+                builder.add(new HttpHeadersBase(header));
+            }
+        }
+
+        return builder.build();
+    }
+
+    final void contentLength(long contentLength) {
+        checkArgument(contentLength >= 0, "contentLength: %s (expected: >= 0)", contentLength);
+        ((HttpHeadersBase) appender()).contentLength(contentLength);
+    }
+
+    @Override
+    public long contentLength() {
+        long contentLength = -1;
+
+        for (StringMultimapGetters<CharSequence, AsciiString> delegate : delegates()) {
+            final HttpHeaderGetters getter = (HttpHeaderGetters) delegate;
+            if (getter.isContentLengthUnknown()) {
+                return -1;
+            }
+
+            final long length = getter.contentLength();
+            if (length < 0) {
+                continue;
+            }
+
+            if (contentLength < 0) {
+                contentLength = 0;
+            }
+            contentLength += length;
```

Comment:
I'm not sure about aggregating all the Content-Length headers as it's unlikely that it accurately represents the actual content length. How about using the length of the first encountered header as we do for others?

Reply:
> How about using the length of the `first` encountered header as we do for others?

✅ updated! agree with this way.

But I'm not sure how we should handle on `isEndOfStream` case 🤔
- e.g. If only some of the parents are `endOfStream: true`, what should we return on `isEndOfStream()`?

PTAL 🙇

Reply:
That's a good question.
Based on previous usage, simply using the flag of the first header may not be sufficient.

I see two potential options:

- If there's a headers with `isEndOfStream`` set to `true``, consider the composite headers as true as well. This approach seems reasonable, as the composite headers should reflect the end-of-stream status if any of them does.
- Ignore the `isEndOfStream` flag in all headers objects and allow the user to specify it after composition.

What do you think of it?
Of course, others might have different opinions. cc @line/dx

Reply:
> If there's a headers with `isEndOfStream` set to `true`, consider the composite headers as `true` as well.

I think `option1` looks better so updated codes~! thank you! 🙇